internal/auth/tokens.go

```diff
@@ -25,65 +27,98 @@ const (
 	GithubJWTAudience = "nscloud.dev/inline-token"
 )
 
+var ErrTokenNotExist = errors.New("token does not exist")
+
+type cachedTokens map[string]Token
+
 type Token struct {
-	TenantToken string `json:"tenant_token,omitempty"`
+	Scopes      []string `json:"scopes,omitempty"`
+	TenantToken string   `json:"tenant_token,omitempty"`
 }
 
-func StoreTenantToken(token string) error {
-	data, err := json.Marshal(Token{TenantToken: token})
+func StoreTenantToken(ctx context.Context, username string, token Token) error {
+	dir, err := dirs.Config()
 	if err != nil {
 		return err
 	}
 
-	configDir, err := dirs.Ensure(dirs.Config())
+	f := filepath.Join(dir, tokenTxt)
+	cache, err := getCachedTokens(ctx, f)
 	if err != nil {
 		return err
 	}
 
-	if err := os.WriteFile(filepath.Join(configDir, tokenTxt), data, 0600); err != nil {
-		return fnerrors.New("failed to write token data: %w", err)
+	sort.Strings(token.Scopes)
+	cache[username+strings.Join(token.Scopes, "")] = token
+
+	data, err := json.Marshal(cache)
+	if err != nil {
+		return err
+	}
+
+	if err := os.WriteFile(f, data, 0600); err != nil {
+		return fnerrors.New("failed to write token cache: %w", err)
 	}
 
 	return nil
 }
 
-func LoadTenantToken(ctx context.Context) (*Token, error) {
+func LoadTenantToken(ctx context.Context, username string, scopes []string) (*Token, error) {
 	dir, err := dirs.Config()
 	if err != nil {
 		return nil, err
 	}
 
-	p := filepath.Join(dir, tokenTxt)
-	data, err := os.ReadFile(p)
+	f := filepath.Join(dir, tokenTxt)
+	cache, err := getCachedTokens(ctx, f)
 	if err != nil {
 		return nil, err
 	}
 
-	// XXX: to invalidate cache if cached token has expired or corrupted - remove it and return [fs.ErrNotExist].
-	cleanTokenCache := func(f string) (*Token, error) {
-		if err := os.Remove(f); err != nil {
-			return nil, err
-		}
-		return nil, fs.ErrNotExist
-	}
-
-	token := &Token{}
-	if err := json.Unmarshal(data, token); err != nil {
-		fmt.Fprintf(console.Debug(ctx), "failed to unmarshal cached tenant token: %v\n", err)
-		return cleanTokenCache(p)
+	sort.Strings(scopes)
+	token, ok := cache[username+strings.Join(scopes, "")]
+	if !ok {
+		return nil, ErrTokenNotExist
 	}
 
 	claims := jwt.RegisteredClaims{}
 	parser := jwt.Parser{}
 	if _, _, err := parser.ParseUnverified(strings.TrimPrefix(token.TenantToken, "nsct_"), &claims); err != nil {
 		fmt.Fprintf(console.Debug(ctx), "failed to parse tenant JWT: %v\n", err)
-		return cleanTokenCache(p)
+		if err := os.Remove(f); err != nil {
+			return nil, err
+		}
+		return nil, ErrTokenNotExist
 	}
 
 	if !claims.VerifyExpiresAt(time.Now(), true) {
```

Comment:
Nit: I find this method name very confusing. From what I read, you are doing the right thing, but would be great to make it more obvious.
How about adding a simple wrapper `IsExpired(claims)`?

Reply:
Added